### DISCUSSION TITLE
How do we handle 404s with domain localization?

### DISCUSSION THREAD

**Discussion by dnish** [amannn/next-intl#638] [Q&A]
Hey,
thank you very much for this awesome library :-). After configuring everything

```
export const config = {
    matcher: ['/((?!api|_next|_vercel|.*\\..*).*)'], // Disable routes for middleware
}

export function middleware(request) {

    const routingObject = {
        locales: ['en','de'],
        defaultLocale:'en',
        localePrefix: 'never',
        localeDetection: false,
        alternateLinks: false,

        domains: [

           {
                domain: 'www.test.com',
                defaultLocale: 'en',
                locales: ['en']
            },

           {
                domain: 'www.test.de',
                defaultLocale: 'de',
                locales: ['de']
            },

        ],

    };

    const handleI18nRouting = createIntlMiddleware(routingObject);
    let response = handleI18nRouting(request);

}

```

This setup works pretty nice, but soon after deploying it for production, I've noticed some errors within our logs:

```
Error: Unable to find `next-intl` locale because the middleware didn't run on this request. See https://next-intl-docs.vercel.app/docs/routing/middleware#unable-to-find-locale
```

This error happens sporadic. After debugging I found out, that urls like https://www.test.com/justan.ico causes this issue. It seems that some browser try to check for a file. In this case, NextJS tries to send a 404 error and calls the root layout. Because we've disabled the middleware for routes with a dot, it throws an error. Now my question is how do we handle this case? My root layout looks like this:

```
export default function RootLayout({ children,params }) {

    const messages = useMessages();

    return (
        <html lang={params.locale} data-theme="main">
        <body>
        <NextIntlClientProvider locale={params.locale} messages={messages}>
            <div className="container mx-auto">
                <Header locale={params.locale}/>
                {children}
                <Footer locale={params.locale}/>
            </div>
        </NextIntlClientProvider>
        </body>
        </html>
    )
}
```

**Comment by amannn** [comment]
Hey, I think you're looking for this: https://next-intl-docs.vercel.app/docs/routing/middleware#matcher-no-prefix

Hope this helps! :)

**Reply by dnish** [reply]
Hey, yeah, that's the page where I got the matcher from above. The issue is that browsers or crawlers may try some unknown file urls, normally we would then reply with a 404 not found, but in this case, we get an internal server error because we ignore the middleware on routes with a dot. So I'm looking for a way to respond with a 404 in such cases.

**Reply by dnish** [reply]
If I call <code>notFound()</code> within my root layout the server error disappears and I get a 404 server response, but on the client side everything will be broken:


```
Warning: validateDOMNesting(...): <div> cannot appear as a child of <#document>.
```


This only happens when the path includes a dot, all other non existing urls are shown with the error component.

**Reply by amannn** [reply]
Ah yep, the routing for 404s when using a `[locale]` folder indeed needs a bit of special care, there are some docs on this: https://next-intl-docs.vercel.app/docs/environments/error-files

Based on your other comment I think you got it working, glad to hear! 🙌

**Comment by dnish** [comment]
Solved the issue, the problem was that I didn't define a root layout within the app folder.


app/layout.js


```
import { Inter, Edu_TAS_Beginner } from 'next/font/google'
import './globals.css'

const inter = Inter({ subsets: ['latin'] });
const EduTASBeginner = Edu_TAS_Beginner({ subsets: ['latin'], variable: '--font-etb', display:'swap'});


export default function RootLayout({ children,params }) {


    return (
        <html lang={params.locale} data-theme="main">
        <body className={`${inter.className} ${EduTASBeginner.variable}`}>
        {children}
        </body>
        </html>
    )
}
```


app/[locale]/layout.js


```
import {Header} from "@/app/components/header/header.js";
import {Footer} from "@/app/components/footer/footer.js";
import {NextIntlClientProvider} from 'next-intl';
import {useMessages} from 'next-intl';
import {notFound} from 'next/navigation';
import {CONFIG} from "@/config.js";


export default function RootLayout({ children,params }) {

    if(!Object.keys(CONFIG.LOCALES).includes(params.locale)) notFound();

    const messages = useMessages();

    return (
        <NextIntlClientProvider locale={params.locale} messages={messages}>
            <div className="container mx-auto">
                <Header locale={params.locale}/>
                {children}
                <Footer locale={params.locale}/>
            </div>
        </NextIntlClientProvider>
    )
}
```

That fixed the problem on client and server side.

**Reply by dnish** [reply]
Okay, this has the downside that we can't set lang={params.locale} because it will be empty within the root directory (app/layout.js). On the other side we have to define a layout.js for handling errors which are not part of the middleware handler 🙁

**Reply by dnish** [reply]
Fixed my code by adding the layout.tsx and not-found.tsx from the demo 😀

https://github.com/amannn/next-intl/tree/main/examples/example-app-router/src/app
